RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+
+The state transition table used to define how the **current state** changes according to the possible **events** is a read only **matrix**. The current state is saved in a writable memory accessible in **kernel space**.
+
+In addition to the verification and monitoring of the system, a monitor can react to an unexpected event. If an **event** is not valid for the **current state**, the automata will report an **unexpected event**. The forms of reaction can vary from logging the event occurrence to the enforcement of the correct behavior to the extreme action of taking a system down to avoid the propagation of a failure.
+
+The [current state ](https://git.kernel.org/pub/scm/linux/kernel/git/bristot/linux.git/tree/include/linux/rv.h#n18)variable is stored in a regular memory space in the kernel. The place varies according to the **monitory type**. It can be either in a **global variable** for the global monitor, on a **per-cpu variable** on a per-cpu monitor, or data stored in the **task struct** for per task monitor.
+
+The same consideration is true for the [monitoring](https://git.kernel.org/pub/scm/linux/kernel/git/bristot/linux.git/tree/include/linux/rv.h#n17) variable; such variable is used to activate or deactivate a specific instance of runtime verification monitor.
+
+On top of such variables (that are defined as part of the da_monitor structure) it is possible to define any custom monitor specific data set that can be associated with the monitor states and be checked against event specific data to make sure the event itself is valid and the data set coming with it is also valid when compared against the current state specific data set.  
+
+
+# Freedom From Interference between RV Monitors and the rest of Kernel code
+
+From an interference perspective we have different types of interference:
+
+
+
+* **Temporal**: Kernel code slowing down the monitor code so much to miss the associated FDTI deadline
+* **Communication**: Kernel code interfering with the monitor code through the interfaces between the Kernel and the RV monitors
+* **Spatial**: Kernel code interfering with the monitor code through random corruption of the Kernel address space
+
+From a **temporal interference** point of view we can assume the safety relevant code (including the Kernel code) to be monitored by an external watchdog. So if such code slows down beyond the allocated max FDTI portion, the external watchdog will trigger. \
+Now the monitored code by definition is safety relevant code, and, as explained above, the associated RV monitor executes in the same context directly invoked by the trace interfaces added in the monitored code. \
+So if the monitored code slows down, the RV Monitor code equally slows down and as a  result the external watchdog is not pet on time, with a safe state driven physically by the watchdog.
```

Comment:
This is true only if the monitor is executed as part of the same queue matching the priority of the least prioritised task/event that is still relevant to safety.

Reply:
the RVM code is executed sequentially like a functiol call departing from the monitored code.....

Reply:
ok, then it falls within the "Unless ..." part of my previous comment, and I'd suggest to incorporate in the text what you just answered to me

Reply:
I can rephrase
<<Now the monitored code by definition is safety relevant code, and, as explained above, the associated RV monitor executes in the same context directly invoked by the trace interfaces added in the monitored code>> 
AS
<<Now the monitored code by definition is safety relevant code, and, as explained above, the associated RV monitor executes sequentially in the same context of the code being monitored, since it is directly invoked by the trace interfaces added in the monitored code>>
If it clarifies better....

Reply:
ok